### PR TITLE
EID-1977 Validation of missing namespace uri on Assertion

### PR DESCRIPTION
**Pull request by idrop**
## What
Given an Assertion AttributeValue missing its QName namespace uri, use the matching namespace uri declared in the parent Response element to build a new QName containing a uri.

A match is considered when the namespace prefix of the Assertion and Response match e.g. they both have value of `eidas-natural`.

The new derived QName is then evaluated against a set of expected AttributeValue QNames.

The parent Response object need to be carried through a series of classes in order to be available on the `EidasAttributeStatementAssertionValidator`.

## How to review
Review `EidasAttributeStatementAssertionValidator` first, and debug the new unit test on `EidasAttributeStatementAssertionValidatorTest`, which contains an eidas Response in the desired state.

## Why
We have an issue validating the qualified name (QName) of decrypted AttributeValue received from the EU Validator Proxy Node, as they do not contain the namespace uri, which is required for evaluation against a set of expected QNames.

For example, given the AttributeValue below:
````
<saml2:Assertion>
......
    <saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
        xsi:type="eidas-natural:CurrentFamilyNameType">...
......
````
There is no definition for `eidas-natural` within the Assertion document.

The parent SAML Response XML document *does* contain the namespace uri i.e.
````
  <saml2p:Response .... 
      xmlns:eidas-natural="http://eidas.europa.eu/attributes/naturalperson" 
````
When evaluating assertions, we are only checking the `<Assertion>` SAML document, and have no reference to the namespace uri defined on the outer `<Response>` document.

### CODE REVIEW

File: hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java [hunk 1]
```diff
@@ -88,6 +90,26 @@ public void shouldThrowIfAttributeValueHasInvalidSchemaType() {
...
+    public void shouldThrowIfAttributeValueHasNullSchemaType() {
+        XMLObject xmlObject = mock(XMLObject.class);
+        when(firstName.getAttributeValues()).thenReturn(List.of(xmlObject));
+        when(xmlObject.getSchemaType()).thenReturn(null);
+
+        validator.validate(assertion);
+    }
+
+    @Test(expected = SamlTransformationErrorException.class)
+    public void shouldNotThrowIfAttributeValueHasNullSchemaNamespace() {
```
Comment (CharlesIC) [inline hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java:103]: `@Test(expected = ...)` indicates the test should throw an exception, but the name suggests it shouldn't (`shouldNotThrow...`). Which is it?

File: hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java [hunk 2]
```diff
@@ -88,6 +90,26 @@ public void shouldThrowIfAttributeValueHasInvalidSchemaType() {
...
+        XMLObject xmlObject = mock(XMLObject.class);
+        when(firstName.getAttributeValues()).thenReturn(List.of(xmlObject));
+        when(xmlObject.getSchemaType()).thenReturn(new QName(null, CurrentGivenName.TYPE_LOCAL_NAME, IdaConstants.EIDAS_NATURUAL_PREFIX));
+        when(firstName.getNameFormat()).thenReturn("urn:oasis:names:tc:SAML:2.0:attrname-format:uri");
```
Comment (CharlesIC) [inline hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java:107]: There's a SAML constant called `URI_REFERENCE` for `"urn:oasis:names:tc:SAML:2.0:attrname-format:uri"` in `org.opensaml.saml.saml2.core.Attribute` class

File: hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java [hunk 3]
```diff
@@ -95,9 +95,11 @@ private void validateAttributes(Assertion assertion) {
                 SamlValidationSpecificationFailure failure = SamlTransformationErrorFactory.emptyAttribute(attributeName);
                 throw new SamlTransformationErrorException(failure.getErrorMessage(), failure.getLogLevel());
             }
-            if (!VALID_TYPE_FOR_ATTRIBUTE.get(attributeName).equals(attribute.getAttributeValues().get(0).getSchemaType())) {
-                final QName schemaType = attribute.getAttributeValues().get(0).getSchemaType();
-                SamlValidationSpecificationFailure failure = SamlTransformationErrorFactory.attributeWithIncorrectType(attributeName, VALID_TYPE_FOR_ATTRIBUTE.get(attributeName), schemaType);
+            QName expectedQName = VALID_TYPE_FOR_ATTRIBUTE.get(attributeName);
+            QName attributeValueQName = attribute.getAttributeValues().get(0).getSchemaType();
+            if (!isExpectedQName(expectedQName, attributeValueQName)) {
+                final QName schemaType = attributeValueQName;
```
Comment (CharlesIC) [inline hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java:101]:
```suggestion

```
This line is not needed - use `attributeValueQName` directly

File: hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java [hunk 4]
```diff
@@ -106,4 +108,10 @@ private void validateAttributes(Assertion assertion) {
             }
         }
     }
+
+    private boolean isExpectedQName(QName expectedQName, QName attributeValueQName) {
+        return attributeValueQName != null
+                && expectedQName.getPrefix().equals(attributeValueQName.getPrefix())
+                && expectedQName.getLocalPart().equals(attributeValueQName.getLocalPart());
+    }
```
Comment (CharlesIC) [inline hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java:116]: This doesn't seem right to me. In this implementation we're not checking the attribute namespace, therefore, we're not checking that the attribute complies with the eIDAS schema.
I believe it's a standard XML practice to define namespaces at the top and then referring to them in children elements - so they can be referred to multiple times without re-defining each time. We should either:
- check if there's a ready XML/OpenSAML method to verify attribute namespaces
- create a map of namespaces in the in `Response` element and then verify the attribute namespace either exists in the attribute or in the map. E.g. for `xsi:type="eidas-natural:CurrentFamilyNameType"` the map should have an entry for key `eidas-natural`

I would check the XML/SAML/eIDAS spec to check where it's allowed to define namespaces to make sure we're looking in the right place.

File: hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java [hunk 5]
```diff
@@ -106,4 +110,21 @@ private void validateAttributes(Assertion assertion) {
...
     }
+
+    private QName getAttributeValueSchemaType(Response response, Attribute attribute) {
+        final QName attributeValueSchemaType = attribute.getAttributeValues().get(0).getSchemaType();
+        if (attributeValueSchemaType.getNamespaceURI().isEmpty()) {
+            Optional<Namespace> matchingPrefix = response.getNamespaces()
+                    .stream()
+                    .filter(ns -> Objects.equals(ns.getNamespacePrefix(), attributeValueSchemaType.getPrefix()))
+                    .findFirst();
+            if (matchingPrefix.isPresent()) {
```
Comment (CharlesIC) [inline hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java:121]: ` .findFirst()` returns Optional so you can chain `ifPresent` directly on the result - no need for the second `if` statement

File: hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java [hunk 6]
```diff
@@ -95,7 +106,20 @@ public void shouldInsistOnMandatoryAttributes() {
...
-        validator.validate(assertion);
+        validator.validate(response, assertion);
+    }
+
+    @Test
+    public void validateEidasAttributesWhenEidasXMLNamespaceURIOnResponse() throws UnmarshallingException, XMLParserException {
+        Security.addProvider(new BouncyCastleProvider());
+        IdaSamlBootstrap.bootstrap();
+        InputStream responseIS = getClass().getClassLoader().getResourceAsStream("response-with-eidas-namespace.xml");
+        InputStream assertionIS = getClass().getClassLoader().getResourceAsStream("assertion-without-eidas-namespace.xml");
```
Comment (CharlesIC) [inline hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java:117]: We don't need XML resource files here - XML objects can be mocked. Other tests in this class do it so the approach can be copied in this test too

File: hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java [hunk 7]
```diff
@@ -106,4 +110,21 @@ private void validateAttributes(Assertion assertion) {
             }
         }
     }
+
+    private QName getAttributeValueSchemaType(Response response, Attribute attribute) {
+        final QName attributeValueSchemaType = attribute.getAttributeValues().get(0).getSchemaType();
+        if (attributeValueSchemaType.getNamespaceURI().isEmpty()) {
+            Optional<Namespace> matchingPrefix = response.getNamespaces()
+                    .stream()
+                    .filter(ns -> Objects.equals(ns.getNamespacePrefix(), attributeValueSchemaType.getPrefix()))
```
Comment (CharlesIC) [inline hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java:119]: Why `Objects.equals` here? I'm suggesting this is the wrong method to use for comparison here, just curious why you've chosen it

File: hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java (hunk 5 above)
Comment (idrop) [reply hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java]: I've now used Optional.map which is similar to ifPresent but allows a return

File: hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java (hunk 7 above)
Comment (idrop) [reply hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java:119]: saves an extra NPE sanity check on `ns.getNamespacePrefix()`

File: hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java (hunk 6 above)
Comment (idrop) [reply hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java]: taken out now 👍  although I think using Mockito is perhaps more cognitive overhead than necessary here

File: hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java [hunk 8]
```diff
@@ -95,15 +108,46 @@ public void shouldInsistOnMandatoryAttributes() {
 
         when(attributeStatement.getAttributes()).thenReturn(List.of(firstName, lastName, dateOfBirth));
 
-        validator.validate(assertion);
+        validator.validate(response, assertion);
+    }
+
+    @Test
+    public void validateCorrectNamespaceURIOnResponseAllowsValidMatchOnAttribute() {
+
```
Comment (CharlesIC) [inline hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java:116]:
```suggestion
```
Seems like unnecessary whitespace?

File: hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java [hunk 9]
```diff
@@ -95,15 +108,46 @@ public void shouldInsistOnMandatoryAttributes() {
...
+        validator.validate(response, assertion);
+
+        verify(response, times(1)).getNamespaces();
+
     }
 
+    @Test
+    public void validateIncorrectNamespaceURIOnResponseWhenNoneOnAttributeValueThrowsException() {
+
```
Comment (CharlesIC) [inline hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidatorTest.java:131]:
```suggestion
```

File: hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java [hunk 10]
```diff
@@ -106,4 +110,19 @@ private void validateAttributes(Assertion assertion) {
...
+        final QName attributeValueSchemaType = attribute.getAttributeValues().get(0).getSchemaType();
+        if (XMLConstants.NULL_NS_URI.equals(attributeValueSchemaType.getNamespaceURI())) {
+            return response.getNamespaces().stream()
+                    .filter(ns -> Objects.equals(ns.getNamespacePrefix(), attributeValueSchemaType.getPrefix()))
+                    .findFirst()
+                    .map(ns -> new QName(
+                            ns.getNamespaceURI(),
+                            attributeValueSchemaType.getLocalPart(),
+                            attributeValueSchemaType.getPrefix()))
+                    .orElse(attributeValueSchemaType);
```
Comment (CharlesIC) [inline hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java:124]: Nice 👌

Comment (instantiator) [reply hub/saml-engine/src/main/java/uk/gov/ida/hub/samlengine/validation/country/EidasAttributeStatementAssertionValidator.java:124]: This is magical. ✨
